### PR TITLE
Fix bug at ssh.py:process() - empty argv[0] Error (#2217) 

### LINKED ISSUE #2217: ssh.process -- Cannot provide empty argv
The following code fails:
```python
from pwn import *
s = ssh("bandit0", "bandit.labs.overthewire.org", 2220, "bandit0")
io = s.process([""], executable="/bin/sh")
io.interactive()
```
```
Traceback (most recent call last):
  File "<string>", line 110, in <module>
ValueError: execve: argv first element cannot be empty
```
We expect to open the process, but instead we get an error.

## Debugging
The culprit is `ssh.py:1034`
```
os.execve(exe, argv, env)
```

`os.execve` disallows empty argv[0]. 

## Proposed solution
Instead of using `os.execve`, we can use `ctypes` to call the `execve` function from libc. Below is a POC to call execve in python2 (python3 works the same but strings need to be replaced with bytes)

```python
import ctypes
# All strings need to be replaced to bytes in python3
exe = "/bin/sh"
argv = []
envp = {"A":"A"}

def get_string_list(string_list):
    #Transform a list of bytes into a ctypes array of char pointers
    char_p_array = (ctypes.c_char_p * len(string_list))()
    for i, string in enumerate(string_list):
        char_p_array[i] = ctypes.c_char_p(string)

    return char_p_array

# Transform envp from dict to list
envp = [k + "=" + v for k,v in envp.items()]

c_exe = ctypes.c_char_p(exe)
c_argv = get_string_list(argv)
c_envp = get_string_list(envp)

# Call execve
libc = ctypes.CDLL(None)
libc.execve(c_exe, c_argv, c_envp)
```

### PR DESCRIPTION
**Pull request by goreil**
Reopening #2225, since the Coverage tests didn't seem to handle changing the target branch correctly.

This addresses #2217 , a bug which disallowed us to use `ssh.py:process()` with an empty argv[0]

# Bug description
`ssh.py.process()` uses a python script to call `os.execve` to create a new process. `os.execve` disallows an empty argv.

# Fix
 This commit replaces `os.execve` with  `ctypes.CDLL(None).execve` to call the C-Library function directly which allows an empty argv.

### CODE REVIEW

**Comment by goreil** [comment]
Hi, just wanted to ask if there are any changes you require me to make to the pull request or is it ready to merge?

**Comment by Arusekk** [comment]
For now I can only tell I am 90% sure that `environb` will break under py2.  I should find some time for a deeper review soon.

**Comment by goreil** [comment]
> For now I can only tell I am 90% sure that `environb` will break under py2. I should find some time for a deeper review soon.

Wow, totally missed that. Thanks for the heads up! Fixed.

**Comment by peace-maker** [comment]
https://github.com/Gallopsled/pwntools/pull/2225#issuecomment-1635984358

The code checks the environment variable name not the value.

File: pwnlib/tubes/ssh.py [hunk 1]
```diff
@@ -1031,7 +1048,29 @@ def is_exe(path):
 %(func_src)s
 %(func_name)s(*%(func_args)r)
 
-os.execve(exe, argv, env)
```
Comment (peace-maker) [inline pwnlib/tubes/ssh.py:1037]: Can we only use the ctypes workaround if `exe != argv[0]`?

File: pwnlib/tubes/ssh.py [hunk 2]
```diff
@@ -1031,7 +1048,29 @@ def is_exe(path):
...
+# os.execve does not allow us to pass empty argv[0]
+# Therefore we use ctypes to call execve directly
+# os.execve(exe, argv, env)
+        
+# Transform envp from dict to list
+env_list = [key + b"=" + value for key, value in env.items()]
```
Comment (peace-maker) [inline pwnlib/tubes/ssh.py:1064]: Please [add a check](https://github.com/python/cpython/blob/025995feadaeebeef5d808f2564f0fd65b704ea5/Modules/posixmodule.c#L6476) that `key` does not contain `=`.

File: pwnlib/tubes/ssh.py [hunk 3]
```diff
@@ -1031,7 +1048,29 @@ def is_exe(path):
...
+# ctypes helper to convert a python list to a NULL-terminated C array
+def to_carray(py_list):
+    py_list += [None] # NULL-terminated
+    return (ctypes.c_char_p * len(py_list))(*py_list)
+
+c_argv = to_carray(argv)
+c_env = to_carray(env_list)
+
+# Call execve
+libc = ctypes.CDLL(None)
```
Comment (peace-maker) [inline pwnlib/tubes/ssh.py]: Do you have a reference of where `ctypes.CDLL` accepts `None` as `name` to get libc? Above we're using `ctypes.CDLL('libc.so.6')`.

**Review by peace-maker — CHANGES_REQUESTED** [review]

**Review by peace-maker — APPROVED** [review]

File: pwnlib/tubes/ssh.py [hunk 4]
```diff
@@ -903,6 +920,9 @@ def process(self, argv=None, executable=None, tty=True, cwd=None, env=None, time
             self.error("executable / argv[0] must be a string: %r" % executable)
         executable = bytearray(packing._need_bytes(executable, min_wrong=0x80))
 
+        # Validate that all key of env do not contain '='
```
Comment (peace-maker) [inline pwnlib/tubes/ssh.py]: This comment seems like a left over.

**Review by peace-maker — CHANGES_REQUESTED** [review]

**Review by peace-maker — APPROVED** [review]

**Comment by peace-maker** [comment]
Thank you! This'll make e.g. overthewire solve scripts cleaner 🔥

**Pull request merged**